src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -2176,71 +2007,237 @@ lExit: ;
+    if (fUnwrapNullable && Nullable::IsNullableType(pMT))
+    {
+        pMT = pMT->GetInstantiation()[0].GetMethodTable();
+    }
 
 #ifdef FEATURE_COMINTEROP
-    // Also do not allow allocation of uninitialized RCWs (COM objects).
+    // COM allocation can involve the __ComObject base type (with attached CLSID) or a
+    // VM-implemented [ComImport] class. For GetUninitializedObject, we block all COM types.
+    // For CreateInstance, the flowchart is:
+    //   - For __ComObject,
+    //     .. on Windows, bypass normal newobj logic and use ComClassFactory::CreateInstance.
+    //     .. on non-Windows, treat as a normal class, type has no special handling in VM.
+    //   - For [ComImport] class, treat as a normal class. VM will replace default
+    //     ctor with COM activation logic on supported platforms, else ctor itself will PNSE.
     if (pMT->IsComObjectType())
-        COMPlusThrow(kNotSupportedException, W("NotSupported_ManagedActivation"));
+    {
+        if (fGetUninitializedObject)
+        {
+            COMPlusThrow(kNotSupportedException, W("NotSupported_ManagedActivation"));
+        }
+
+        if (IsComObjectClass(typeHandle))
+        {
+#ifdef FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+            // Need to enter cooperative mode to manipulate OBJECTREFs
+            GCX_COOP();
+            SyncBlock* pSyncBlock = pRuntimeType.Get()->GetSyncBlock();
+            pClassFactory = (void*)pSyncBlock->GetInteropInfo()->GetComClassFactory();
+#endif // FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+
+            if (!pClassFactory)
+            {
+                COMPlusThrow(kInvalidComObjectException, IDS_EE_NO_BACKING_CLASS_FACTORY);
+            }
+
+            fRequiresSpecialComActivationStub = true;
+        }
+    }
 #endif // FEATURE_COMINTEROP
 
-    // If it is a nullable, return the underlying type instead.
-    if (Nullable::IsNullableType(pMT))
-        pMT = pMT->GetInstantiation()[0].GetMethodTable();
+    pMT->EnsureInstanceActive();
 
-    retVal = pMT->Allocate();
+    // All checks passed! Pass parameters back to the caller.
+
+    if (fRequiresSpecialComActivationStub)
+    {
+        // managed sig: ComClassFactory* -> object (via FCALL)
+        // caller understands the -1 sentinel value and substitutes real address
+        *ppfnAllocator = (PCODE)(-1);
+        *pvAllocatorFirstArg = pClassFactory;
+    }
+    else
+    {
+        // managed sig: MethodTable* -> object (via JIT helper)
+        *ppfnAllocator = CEEJitInfo::getHelperFtnStatic(CEEInfo::getNewHelperStatic(pMT));
+        *pvAllocatorFirstArg = pMT;
+
+        if (ppfnCtor != NULL && pMT->HasDefaultConstructor())
+        {
+            // managed sig: object -> void
+            // for ctors on value types, lookup boxed entry point stub
+            MethodDesc* pMD = pMT->GetDefaultConstructor(pMT->IsValueType());
+            _ASSERTE(pMD != NULL);
+
+            pMD->EnsureActive();
```

Comment:
This is not necessary since you have done `EnsureInstanceActive` on the MethodTable above